contributors/design-proposals/public-identities-for-pods.md

```diff
@@ -0,0 +1,132 @@
+# Public identities for Pods
+
+* Authors: Sebastian Łaskawiec (@slaskawi)
+* Last edit: [2017-03-13](#history)
+* Status: design
+
+Approvers:
+* [ ] smarterclayton
+* [ ] bprashanth
+* [ ] enisoc
+* [ ] foxish
+* [ ] janetkuo
+* [ ] kargakis
+* [ ] kow3ns
+
+**Table of Contents**
+
+* [Goals](#goals)
+* [Non-goals](#non-goals)
+* [API](#api)
+* [Behavior](#behavior)
+* [Upgrading](#upgrading)
+* [Implementation](#implementation)
+* [Alternatives](#alternatives)
+* [History](#history)
+
+# Goals
+
+* The main goal of Public Identities for Pods is to expose each Pod as an external IP address.
+  This approach allows sophisticated applications to communicate with the outside world using custom protocols.
+  A examples which require this approach are:
+  * Game Servers (which often use RTP, UDP)
+  * Data Grids (with L7 routing, such as [Infinispan](http://infinispan.org/docs/dev/user_guide/user_guide.html#using_hot_rod_server))
+  * Media Servers
+
+* The secondary goal is to create an API similar to `LoadBalancer` to manage external IP addresses.
+
+* The implementation is targeted (but not limited to) StatefulSets. Stable hostnames allow to identify the same
+  Stateful Pod after crash or restart (it should have a stable hostname). However it also makes sense to expose
+  ordinary deployments this way.
+
+# Existing solution
+
+* The above goal is partially possible to achieve with Loadbalanced Services, NodePort Services or
+  Ingress with session stickiness (if HTTP is the main protocol used for communication) but very often
+  this is not good enough and application clients assume there is only one application instance
+  behind an IP address (in other words there 1 to 1 mapping between IP addresses and Application
+  instances). Therefore a specialized implementation exposing an IP address per Pod is required.
+
+# API
+
+A new type of Service needs to be created in order to indicate that an external IP address will be exposed per Pod:
+
+```go
+const (
+    //Existing Service types
+	ServiceTypeClusterIP ServiceType = "ClusterIP"
+	ServiceTypeNodePort ServiceType = "NodePort"
+	ServiceTypeLoadBalancer ServiceType = "LoadBalancer"
+	ServiceTypeExternalName ServiceType = "ExternalName"
+
+	//new Service type - expose IP per Pod
+	ServiceTypeIPPerPod ServiceType = "IPPerPod"
+)
+```
+
+A new interface in `cloud.go` will be created to allow creating an external IP addresses.
```

Comment:
@kow3ns mentioned in the comment below about Elastic IP (AWS) and Static External IPs (GCP). Is it possible to use those things somehow?